OpenRA.Game/Map/IMap.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Collections.Generic;
+using OpenRA.Primitives;
+using OpenRA.Support;
+using OpenRA.Traits;
+
+namespace OpenRA
+{
+	public interface IMap
+	{
+		Dictionary<CPos, TerrainTile> ReplacedInvalidTerrainTiles { get; }
+		MapGrid Grid { get; }
+		int2 MapSize { get; }
+		string Tileset { get; }
+		Ruleset Rules { get; }
+		CellRegion AllCells { get; }
+		List<CPos> AllEdgeCells { get; }
+		WDist CellHeightStep { get; }
+		CellLayer<byte> CustomTerrain { get; }
+		WPos ProjectedBottomRight { get; }
+		PPos[] ProjectedCells { get; }
+		WPos ProjectedTopLeft { get; }
+		Rectangle Bounds { get; }
+		int MapFormat { get; }
+		byte GetTerrainIndex(CPos cell);
+		TerrainTypeInfo GetTerrainInfo(CPos cell);
+		void NewSize(Size size, ITerrainInfo terrainInfo);
+		WVec Offset(CVec delta, int dz);
+		WAngle FacingBetween(CPos cell, CPos towards, WAngle fallbackfacing);
+		IEnumerable<CPos> FindTilesInAnnulus(CPos center, int minRange, int maxRange, bool allowOutsideBounds = false);
+		IEnumerable<CPos> FindTilesInCircle(CPos center, int maxRange, bool allowOutsideBounds = false);
+		CPos CellContaining(WPos pos);
+		PPos ProjectedCellCovering(WPos pos);
+		WDist DistanceAboveTerrain(WPos pos);
+		WPos CenterOfCell(CPos cell);
+		WPos CenterOfSubCell(CPos cell, SubCell subCell);
+		CPos ChooseClosestEdgeCell(CPos cell);
+		MPos ChooseClosestEdgeCell(MPos uv);
+		CPos ChooseClosestMatchingEdgeCell(CPos cell, Func<CPos, bool> match);
+		CPos ChooseRandomCell(MersenneTwister rand);
+		CPos ChooseRandomEdgeCell(MersenneTwister rand);
+		CPos Clamp(CPos cell);
+		MPos Clamp(MPos uv);
+		PPos Clamp(PPos puv);
+		bool Contains(CPos cell);
+		bool Contains(MPos uv);
+		bool Contains(PPos puv);
+		WDist DistanceToEdge(WPos pos, in WVec dir);
+		PPos[] ProjectedCellsCovering(MPos uv);
+
+		event Action<CPos> CellProjectionChanged;
+		byte ProjectedHeight(PPos puv);
+		void Resize(int width, int height);
+		void SetBounds(PPos tl, PPos br);
+		WRot TerrainOrientation(CPos cell);
+		List<MPos> Unproject(PPos puv);
+		(Color Left, Color Right) GetTerrainColorPair(MPos uv);
+		byte[] SavePreview();
+	}
+
+	public interface IMapElevation : IMap
```

Comment:
code above does not check if a map implements this interface. does a direct cast. likely because also maps without elevation implement it(?).

may as well add this to IMap.

Reply:
Not sure if you referenced the right code above as I don't see the cast. However the idea is that functions that are not implemented would not be called in the first place, because the mod would not call them. E.g.: Elevation isn't used in Red Alert, so it would never call functions that have casts to `IMapElevation`.